### PR TITLE
Basic to poly

### PR DESCRIPTION
**Pull request by srajangarg**
This is a more generic `find_gen` implementation (and also much much cleaner :stuck_out_tongue:), which uses two visitors. One is the normal one, while the other one deals with stuff inside a Pow. For eg, `(x/2 + 1)` just like that has generator `x`, but `2**(x/2 + 1)` has generator `2**(x/2)` and not `2**x` (if we were to just call `find_gen` recursively). So special rules apply when calling the `find_gen` recursively inside powers.

### CODE REVIEW

**Comment by srajangarg** [comment]
In the normal visitor, things are stored in the `pow` fashion  (in `gen_set`), The corresponding powers must always be positive. For eg.
`x` -> `{x, 1}`
`x**-1` -> `{x**-1, 1}`
`x**(3/2)` -> `{x, (3/2)}` (thought this can never be a generator, as we choose `{x, 1/2}`)
`2**x` -> `{2**x, 1}` and so on.

The pow visitor stores the coefficients of a `Basic` (like in Add) (in `gen_set`), with the same rule, that the numbers must be positive.
`-x` -> `{-x, 1}`
`-3x/2` -> `{-x, 3/2}` (but this can never be a generator, as we choose `{-x, 1/2}`)

Thus, `gen_set` of a `Pow` = `{pow(Pow.base, it.first), it.second}` for `it` in `gen_set` of `Pow.exp`.
Sorry, if the idea is not very succinctly explained.

**Comment by certik** [comment]
Thanks for working on this. Looks good to me.

**Comment by srajangarg** [comment]
Work on `find_gen` is done from my side, I would really appreciate it if I can get more test cases from others. I will start on basic->uintpoly conversion now.

**Comment by certik** [comment]
Can you please fix the formatting, per https://travis-ci.org/symengine/symengine/jobs/141314710

**Comment by srajangarg** [comment]
Yes, will fix it as I push more commits in. There's still work left.

**Comment by certik** [comment]
Ok. Ping us once you want this reviewed.

**Comment by srajangarg** [comment]
@certik @isuruf ping

**Comment by srajangarg** [comment]
I was just testing out how the parser is working out. It is working very seamlessly, constructing polynomials has never been easier :smile:
```
s = "2*(x+1)**10 + 3*(x+2)**5";
poly = UIntPoly::from_basic(parse(s));

s = "((x+1)**5)*(x+2)*(2*x + 1)**3";
poly1 = UIntPoly::from_basic(parse(s));
```

**Comment by Sumith1896** [comment]
Great job!

File: symengine/polys/basic_to_poly.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        b.accept(*this);
+        return gen_set;
+    }
+
+    void bvisit(const Pow &x)
+    {
+        umap_basic_num pow_pairs
+            = _find_gens_poly_pow(x.get_exp(), x.get_base());
+        for (auto it : pow_pairs)
+            gen_set[pow(x.get_base(), it.first)] = it.second;
```
Comment (isuruf) [inline symengine/polys/basic_to_poly.cpp]: I don't understand what this method does. Can you give an example?

Comment (srajangarg) [inline symengine/polys/basic_to_poly.cpp]: Take `2**(3*x + x**2 + (y/2))`
`find_gen(2**(3*x + x**2 + (y/2))`
`= 2** find_gen_pow(3*x + x**2 + (y/2))`
`= 2** (x, x**2, y/2)`
`= (2**x, 2**x**2, 2**(y/2))`

So, in  a way `find_gen_pow` does with coefficients what `find_gen` does with exponents. As in `find_gen` the exponents are stored for each generator, and the exponents is always of the type `1/d` where `d` belongs to natural numbers, similarly for `find_gen_pow` the coefficients are always stored in `1/d` form.

Thus like in `find_gen` we can't have `x` and `x**1/2` both, because one is square of other, we can't have `x` and `x/2` both in `find_gen_pow` as in exponent terms, one is square of the other.

Comment (isuruf) [inline symengine/polys/basic_to_poly.cpp]: I understand how `find_gen_pow` works, but what I'm confused about is why in `2**(x**2)` you return `2**x` as the generator?

Comment (srajangarg) [inline symengine/polys/basic_to_poly.cpp]: I don't understand, I will return `2**(x**2)` as the generator.
`find_gen(2**(x**2)`
`= 2** find_gen_pow(x**2)`
`= 2** {{x**2 : 1}}`
`= 2**(x**2)`

Comment (isuruf) [inline symengine/polys/basic_to_poly.cpp]: Sorry that was a poor example. Can you explain what happens to `2**(x**(a+1))`? What I'm trying to say is that you don't need the recursion here.

Comment (srajangarg) [inline symengine/polys/basic_to_poly.cpp]: I think we do consider, `2**(2**(x+1))`
If no recursion is used we get `gen = 2**(2**(x+1))`, while the actual generator is `2**2**x`

Comment (srajangarg) [inline symengine/polys/basic_to_poly.cpp]: Without recursion
`find_gen(2**(2**(x+1)))`
`= 2 ** find_gen_pow(2**(x+1))`
`= 2 ** {{2**(x+1) : 1}}`
`= 2 ** (2**(x+1))`

With recursion
`find_gen(2**(2**(x+1)))`
`= 2 ** find_gen_pow(2**(x+1))`
`= 2 ** (2 ** find_gen_pow(x + 1))`
`= 2 ** (2** {{x : 1}})`
`= 2**2**x`

The important point here was that `1` was not returned by `find_gen_pow(x+1)` as it saw that `2**1` is an integer and may very well be a coefficient of this term and not a generator.

Comment (srajangarg) [inline symengine/polys/basic_to_poly.cpp]: Though I see how it fails on `2**(x**(a+1))`
`find_gen(2**(x**(a+1)))`
`= 2 ** find_gen_pow(x**(a+1))`
`= 2 ** (x ** find_gen_pow(a + 1))`
`= 2 ** (x** {{a : 1}, {1 : 1}})`
`= 2**x**(a, 1)`
`= (2**x**a, 2**x)`

File: symengine/polys/basic_conversions.h [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+template <typename T, typename P>
+enable_if_t<std::is_same<T, UIntDict>::value
+#ifdef HAVE_SYMENGINE_PIRANHA
+                or std::is_same<T, pintpoly>::value
+#endif
+#ifdef HAVE_SYMENGINE_FLINT
+                or std::is_same<T, fzp_t>::value
+#endif
+            ,
+            T>
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: How about checking that `P` is a base of `UPolyBase`?

Comment (isuruf) [inline symengine/polys/basic_conversions.h]: I meant `UIntPolyBase`

File: symengine/polys/basic_conversions.h [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+_b_to_upoly(const RCP<const Basic> &basic, const RCP<const Basic> &gen)
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Use something like `basic_to_upoly`

File: symengine/polys/basic_conversions.h [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    D dict;
+
+    D apply(const Basic &b, const RCP<const Basic> &gen_)
+    {
+        gen = gen_;
+        b.accept(*this);
+        return dict;
+    }
+
+    virtual void d_add_term(unsigned int pow, const Basic &x) = 0;
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: You don't have to use virtual functions when you are using CRTP.

File: symengine/polys/basic_conversions.h [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+class BasicToUExprPoly : public BasicToUPolyBase<UExprPoly, BasicToUExprPoly>
+{
+public:
+    using BasicToUPolyBase<UExprPoly, BasicToUExprPoly>::bvisit;
+    using BasicToUPolyBase<UExprPoly, BasicToUExprPoly>::apply;
+
+    void bvisit(const Rational &x)
+    {
+        if (not x.is_zero())
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: This is always true

File: symengine/polys/basic_conversions.h [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+{
+    RCP<const Basic> exp = basic;
+    if (ex)
+        exp = expand(basic);
+    return P::from_container(
+        gen, _b_to_upoly<typename P::container_type, P>(exp, gen));
+}
+
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic, bool ex = false)
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Can you copy this definition to the top of the file and keep the implementation here? It'll be easier to read about what the header is providing.

File: symengine/series_flint.h [hunk 7]
```diff
@@ -10,12 +10,13 @@
 namespace SymEngine
 {
 
-using fp_t = fmpq_poly_wrapper;
+using fqp_t = fmpq_poly_wrapper;
```
Comment (isuruf) [inline symengine/series_flint.h]: Why this change?

Comment (srajangarg) [inline symengine/series_flint.h]: In two places there was
`using fp_t = fmpq_poly_wrapper`
`using fp_t = fmpz_poly_wrapper`

It was causing issues, as I needed to use the `fmpz` version but the `fmpq` one was defined later (in the includes). So changed it to end the ambiguity.

File: symengine/polys/basic_conversions.h (hunk 2 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: How can I do that?

Comment (isuruf) [inline symengine/polys/basic_conversions.h]: http://en.cppreference.com/w/cpp/types/is_base_of

File: symengine/polys/basic_conversions.h [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            d_add_term(powr, *coef);
+
+        } else {
+            d_add_term(0, x);
+        }
+    }
+
+    void bvisit(const Add &x)
+    {
+        dict = _basic_to_upoly<D, P>(x.coef_, gen);
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Reuse the visitor here

File: symengine/polys/basic_conversions.h [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        for (auto const &it : x.dict_)
+            dict += (_basic_to_upoly<D, P>(it.first, gen)
+                     * _basic_to_upoly<D, P>(it.second, gen));
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Same here. Don't use `_basic_to_upoly` here

File: symengine/polys/basic_conversions.h [hunk 10]
```diff
@@ -0,0 +1,221 @@
+#include <symengine/visitor.h>
+
+namespace SymEngine
+{
+
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic,
+                        const RCP<const Basic> &gen, bool ex = false);
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic, bool ex = false);
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Can you add some documentation about these two methods? (A comment above each method would do)

File: symengine/expression.h [hunk 11]
```diff
@@ -173,11 +173,6 @@ inline Expression expand(const Expression &arg)
     return expand(arg.get_basic());
 }
 
-inline bool unified_eq(const Expression &a, const Expression &b)
```
Comment (isuruf) [inline symengine/expression.h]: Why this change?

File: symengine/series_flint.h (hunk 7 above)
Comment (isuruf) [inline symengine/series_flint.h]: Great. Would have been better if it was a new PR.

File: symengine/polys/basic_conversions.h [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    using BasicToUPolyBase<UExprPoly, BasicToUExprPoly>::apply;
+
+    void bvisit(const Rational &x)
+    {
+        dict = UExprDict(x.rcp_from_this());
+    }
+
+    void d_add_term(unsigned int pow, const Basic &x)
+    {
+        dict = UExprDict({{pow, x.rcp_from_this()}});
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: I initially thought `d_add_term` meant adding a term to a dictionary, but it's not. Can you change the name?

File: symengine/polys/basic_conversions.h [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        dict = _basic_to_upoly<D, P>(x.coef_, gen);
+        for (auto const &it : x.dict_)
+            dict *= _basic_to_upoly<D, P>(pow(it.first, it.second), gen);
+    }
+
+    void bvisit(const Integer &x)
+    {
+        integer_class i = x.i;
+        if (i != 0)
+            dict = P::container_from_dict(gen, {{0, i}});
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: What happens when `i == 0` ?

File: symengine/polys/basic_conversions.cpp [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+        }
+
+        if (is_a<const Rational>(*exp) and is_a<const Rational>(*it->second)) {
+            gen_set[base] = rcp_static_cast<const Rational>(div(
+                one,
+                lcm(*rcp_static_cast<const Rational>(exp)->get_den(),
+                    *rcp_static_cast<const Rational>(it->second)->get_den())));
+        } else if (is_a<const Rational>(*exp)) {
+            gen_set[base] = rcp_static_cast<const Rational>(
+                div(one, rcp_static_cast<const Rational>(exp)->get_den()));
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: Use `divnum` here

File: symengine/polys/basic_conversions.cpp [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+                        const RCP<const Number> &exp)
+    {
+        auto it = gen_set.find(base);
+        if (it == gen_set.end()) {
+            gen_set[base] = exp;
+            return;
+        }
+
+        if (is_a<const Rational>(*exp) and is_a<const Rational>(*it->second)) {
+            gen_set[base] = rcp_static_cast<const Rational>(div(
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: Use `divnum` here

File: symengine/polys/basic_conversions.cpp [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    void add_to_gen_set(const RCP<const Basic> &base,
+                        const RCP<const Number> &exp)
+    {
+        auto it = gen_set.find(base);
+        if (it == gen_set.end()) {
+            gen_set[base] = exp;
+            return;
+        }
+
+        if (is_a<const Rational>(*exp) and is_a<const Rational>(*it->second)) {
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: Restructure the code like
```
if (is_a<const Rational>(*exp)) {
    if (is_a<const Rational>(*it->second)) {

   } else {

   }
}

``

File: symengine/polys/basic_conversions.cpp [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+            }
+
+        } else if (is_a<const Rational>(*x.get_exp())) {
+
+            RCP<const Basic> mulx = one;
+            if (rcp_static_cast<const Rational>(x.get_exp())->is_negative())
+                mulx = minus_one;
+            RCP<const Rational> den = rcp_static_cast<const Rational>(div(
+                one, rcp_static_cast<const Rational>(x.get_exp())->get_den()));
+            add_to_gen_set(pow(x.get_base(), mulx), den);
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]:
```
RCP<const Basic> mulx = one;
RCP<const Rational> rat = rcp_static_cast<const Rational>(x.get_exp());
if (r->is_negative())
     mulx = minus_one;
add_to_gen_set(pow(x.get_base(), mulx), divnum(one, rat->get_den()));
```

File: symengine/polys/basic_conversions.cpp [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+            RCP<const Basic> mulx = one, divx = one;
+
+            if (it.second->is_negative())
+                mulx = minus_one;
+
+            if (is_a<const Rational>(*it.second))
+                divx = rcp_static_cast<const Rational>(it.second)->get_den();
+
+            gen_set[mul(mulx, it.first)]
+                = rcp_static_cast<const Number>(div(one, divx));
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: Use `divnum` here

File: symengine/polys/basic_conversions.cpp [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+
+        if (x.coef_->is_negative())
+            mulx = minus_one;
+
+        if (is_a<const Rational>(*x.coef_))
+            divx = rcp_static_cast<const Rational>(x.coef_)->get_den();
+
+        auto dict = x.dict_;
+        gen_set[Mul::from_dict(mulx, std::move(dict))]
+            = rcp_static_cast<const Number>(div(one, divx));
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: Use `divnum`

File: symengine/polys/basic_conversions.cpp [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    void bvisit(const Number &x)
+    {
+        if (not is_a_Number(*pow(the_base, x.rcp_from_this()))) {
+            if (x.is_positive())
+                gen_set[one] = x.rcp_from_this_cast<const Number>();
+            else
+                gen_set[minus_one]
+                    = rcp_static_cast<const Number>(neg(x.rcp_from_this()));
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: Use `mulnum(x.rcp_from_this_cast<const Number>(), minus_one)`

File: symengine/polys/basic_conversions.h [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+                                                            x.get_base(), gen)),
+                                i)
+                          ->get_poly();
+                return;
+            }
+        }
+
+        RCP<const Basic> genbase = gen, genpow = one, coef = one, tmp;
+        if (is_a<const Pow>(*gen)) {
+            genbase = rcp_static_cast<const Pow>(gen)->get_base();
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Since you don't need to store a RCP, you can use `static_cast<const Pow &>(gen).get_base()`

File: symengine/expression.h (hunk 11 above)
Comment (srajangarg) [inline symengine/expression.h]: I was getting an ambiguous function call, in calling `unified_eq(umap_basic_num, umap_basic_num)`. So I moved the `unified_eq` to [this](https://github.com/symengine/symengine/pull/1003/files#diff-2b0bba2525514ed902f57cc760843b54R136)

I think compiler could not figure out which function to call? The expression one or the templatized one?

Comment (isuruf) [inline symengine/expression.h]: Are you sure? Can you send an error log? If there is an error, we should fix that.

File: symengine/polys/basic_conversions.h (hunk 12 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: `d_set_term`?

File: symengine/expression.h (hunk 11 above)
Comment (srajangarg) [inline symengine/expression.h]: Yes, I'm sure. You can just check on master, try calling `unified_eq` on two `umap_basic_num`.

Comment (srajangarg) [inline symengine/expression.h]: Actually you can just call it on two `Numbers` I think.

File: symengine/polys/basic_conversions.h (hunk 21 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: Didn't get this. You mean
`genbase = static_cast<const Pow &>(*gen).get_base()` ? `gen` is already an RCP

Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Yes. `gen` is already an RCP, but `rcp_static_cast<const Pow>` would create a new `RCP` which we don't need.
`genbase = static_cast<const Pow &>(*gen).get_base()` would work.

File: symengine/polys/basic_conversions.h (hunk 13 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: Dictionary remains empty, like it should.

File: symengine/polys/basic_conversions.h (hunk 8 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: Can do `x.coef->accept(*this)` here

File: symengine/polys/basic_conversions.h (hunk 9 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: I don't think there is an alternative to this, what do you mean?

Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Yes, but you can store the resulting `dict`.

File: symengine/polys/basic_conversions.cpp [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+    {
+        auto it = gen_set.find(base);
+        if (it == gen_set.end()) {
+            gen_set[base] = exp;
+            return;
+        }
+
+        if (is_a<const Rational>(*exp)) {
+            RCP<const Integer> den
+                = rcp_static_cast<const Rational>(exp)->get_den();
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: ` rcp_static_cast<const Rational>(exp)->get_den()` -> `static_cast<const Rational &>(*exp).get_den()`

File: symengine/polys/basic_conversions.cpp [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+            if (is_a<const Rational>(*it->second))
+                gen_set[base] = divnum(
+                    one, lcm(*den, *rcp_static_cast<const Rational>(it->second)
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: `static_cast<const Rational &>(*it->second)`

File: symengine/polys/basic_conversions.cpp [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+            x.coef_->accept(*this);
+
+        for (auto it : x.dict_) {
+            RCP<const Number> mulx = one, divx = one;
+
+            if (it.second->is_negative())
+                mulx = minus_one;
+
+            if (is_a<const Rational>(*it.second))
+                divx = rcp_static_cast<const Rational>(it.second)->get_den();
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: Same here

File: symengine/polys/basic_conversions.cpp [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+    {
+        // won't handle cases like 2**((x+1)(x+2))
+        // needs `expand` to have been called
+        RCP<const Number> mulx = one, divx = one;
+
+        if (x.coef_->is_negative())
+            mulx = minus_one;
+
+        if (is_a<const Rational>(*x.coef_))
+            divx = rcp_static_cast<const Rational>(x.coef_)->get_den();
```
Comment (isuruf) [inline symengine/polys/basic_conversions.cpp]: You can do `divx = divnum(one, static_cast<const Rational &>(x*.coef_).get_den())` and modify Line 143

File: symengine/polys/basic_conversions.h [hunk 26]
```diff
@@ -0,0 +1,233 @@
...
+// convert a `basic`, to a UPoly `P` (eg. UIntPoly, UExprPoly, UIntPolyFlint)
+// using `gen` as the genarator. Throws, if poly constructions not possible.
+// `ex` is the optional parameter for epanding the given `basic` or not.
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic,
+                        const RCP<const Basic> &gen, bool ex = false);
+// convert a `basic`, to a UPoly `P` (eg. UIntPoly, UExprPoly, UIntPolyFlint)
+// after finding out the generator automatically. Throws, if number
+// of generators found != 1, or poly construction not possible.
+// `ex` is the optional parameter for epanding the given `basic` or not.
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: `epanding` -> `expanding`

File: symengine/polys/basic_conversions.h (hunk 9 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: Can you write in pseudocode what you mean, I'm not sure I understand.

Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Something like this
```
D res = apply(coef);
for (auto const &it : x.dict_)
    res += apply(it.first) * apply(it.second);
dict = std::move(res);
```

File: symengine/polys/basic_conversions.h [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+public:
+    RCP<const Basic> gen;
+    using D = typename P::container_type;
+    D dict;
+
+    D apply(const Basic &b, const RCP<const Basic> &gen_)
+    {
+        gen = gen_;
+        b.accept(*this);
+        return dict;
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Instead of returning a copy, move it.
